### PR TITLE
Don't panic just fail when cannot connect to any memcached server.

### LINKED ISSUE #24: vcl_init should fail and not panic
See : https://github.com/varnish/libvmod-memcached/issues/18#issuecomment-372597943

### PR DESCRIPTION
**Pull request by dmatetelki**
Fixes #24

### CODE REVIEW

File: src/vmod_memcached.c [hunk 1]
```diff
@@ -76,7 +76,10 @@ get_memcached(const struct vrt_ctx *ctx, struct vmod_mc_vcl_settings *settings)
 	struct timespec wait;
 
 	CHECK_OBJ_NOTNULL(settings, VMOD_MC_SETTINGS_MAGIC);
-	AN(settings->pool);
+	if (!settings->pool) {
+		VSL(SLT_Error, 0, "Could not connect to any memcached servers");
```
Comment (dridi) [inline src/vmod_memcached.c]: In addition to logging the error, you should also fail ~the transaction~ `vcl_init`.

Comment (dmatetelki) [reply src/vmod_memcached.c]: I have to admit I don't know how to do that. I'll read up.

**Comment by dmatetelki** [comment]
@Dridi : It seems to me that making a vcl fail changed :

4.x :
`VSLb(ctx->vsl, SLT_Error, "error");`

6.0 added new:
`VRT_fail(ctx, "error");`

The first one works on both versions. 
I also saw:
```
MSG(ctx, "error");
VRT_handling(ctx, VCL_RET_FAIL);
```
What is the preferred way to make `vcl_init` fail ?

**Comment by dridi** [comment]
`VSLb` + `VRT_handling` should work for both 4.1 and later releases (in essence that's what `VRT_fail` does).

Where did you see the `MSG` bit?

File: src/vmod_memcached.c [hunk 2]
```diff
@@ -76,7 +76,10 @@ get_memcached(const struct vrt_ctx *ctx, struct vmod_mc_vcl_settings *settings)
 	struct timespec wait;
 
 	CHECK_OBJ_NOTNULL(settings, VMOD_MC_SETTINGS_MAGIC);
-	AN(settings->pool);
+	if (!settings->pool) {
+		VSLb(ctx->vsl, SLT_Error, "Could not connect to memcached.");
+		VRT_handling(ctx, VCL_RET_FAIL);
+	}
```
Comment (dridi) [inline src/vmod_memcached.c]: It might be a good idea to also `return (NULL);` and not run into `memcached_pool_fetch` with a `NULL` pool.

**Comment by dridi** [comment]
I'd like to see a test case for the sake of completeness, I will attempt writing one myself.

**Comment by dmatetelki** [comment]
@Dridi : How about this ? Generating a bad `memcached.servers.vcl` by passing a bad PID (1)

```
varnishtest "Test bad vcl"

shell {${vmod_topsrc}/src/tests/gen-vcl.sh 1 "--SERVER=@SERVER@"}

server s1 {
	rxreq
	txresp
} -start

varnish v1 -vcl+backend {
	import memcached from "${vmod_topbuild}/src/.libs/libvmod_memcached.so";

	sub vcl_init {
		include "${tmpdir}/memcached.servers.vcl";
	}

	sub vcl_deliver {
		memcached.set("test01", "Hello, World", 0, 0);
	}
} -start

logexpect l1 -v v1 {
	expect * * Error  "Could not connect to memcached"
} -start

client c1 {
	txreq -url "/"
	rxresp
} -run
```

**Comment by dridi** [comment]
I'm a bit confused because `test07.vtc` seems to cover the no-connection case. Does the test suite still pass with your patch?

**Comment by dmatetelki** [comment]
Yes, my patch runs with all 1-7 tests plus the one i posted as a commit.
Without my patch, none of the 1-7 triggers the problem just the posted.

**Comment by dridi** [comment]
Confirmed on my machine. Add this test to your patch and I'll be happy to merge it and close the last outstanding issue!

**Comment by dmatetelki** [comment]
@Dridi : test added

**Pull request merged**